v2/internal/reconcilers/postgresql/postgresql_user_reconciler.go

```diff
@@ -0,0 +1,247 @@
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/Azure/azure-service-operator/v2/internal/reconcilers"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	postgresqlutil "github.com/Azure/azure-service-operator/v2/internal/util/postgresql"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/conditions"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/core"
+)
+
+var _ genruntime.Reconciler = &PostgreSqlUserReconciler{}
+
+type PostgreSqlUserReconciler struct {
+	reconcilers.ARMOwnedResourceReconcilerCommon
+	ResourceResolver *resolver.Resolver
+	Config           config.Values
+}
+
+func NewPostgreSqlUserReconciler(
+	kubeClient kubeclient.Client,
+	resourceResolver *resolver.Resolver,
+	positiveConditions *conditions.PositiveConditionBuilder,
+	cfg config.Values) *PostgreSqlUserReconciler {
+
+	return &PostgreSqlUserReconciler{
+		ResourceResolver: resourceResolver,
+		Config:           cfg,
+		ARMOwnedResourceReconcilerCommon: reconcilers.ARMOwnedResourceReconcilerCommon{
+			ResourceResolver: resourceResolver,
+			ReconcilerCommon: reconcilers.ReconcilerCommon{
+				KubeClient:         kubeClient,
+				PositiveConditions: positiveConditions,
+			},
+		},
+	}
+}
+
+func (r *PostgreSqlUserReconciler) asUser(obj genruntime.MetaObject) (*asopostgresql.User, error) {
+	typedObj, ok := obj.(*asopostgresql.User)
+	if !ok {
+		return nil, errors.Errorf("cannot modify resource that is not of type *asopostgresql.User. Type is %T", obj)
+	}
+
+	return typedObj, nil
+}
+
+func (r *PostgreSqlUserReconciler) CreateOrUpdate(ctx context.Context, log logr.Logger, eventRecorder record.EventRecorder, obj genruntime.MetaObject) (ctrl.Result, error) {
+	user, err := r.asUser(obj)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Augment Log
+	log = log.WithValues("azureName", user.AzureName())
+
+	// Resolve the secrets
+	secrets, err := r.ResourceResolver.ResolveResourceSecretReferences(ctx, user)
+	if err != nil {
+		return ctrl.Result{}, reconcilers.ClassifyResolverError(err)
+	}
+
+	db, err := r.connectToDB(ctx, log, user, secrets)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	defer db.Close()
+
+	log.V(Status).Info("Creating PostgreSql user")
+
+	password, err := secrets.LookupFromPtr(user.Spec.LocalUser.Password)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to look up .spec.localUser.Password")
+	}
+
+	// Create or update the user. Note that this updates password if it has changed
+	username := user.Spec.AzureName
+	err = postgresqlutil.CreateOrUpdateUser(ctx, db, user.Spec.AzureName, password)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to create user")
+	}
+
+	// Ensure that the privileges are set
+	err = postgresqlutil.ReconcileGrantUserRoles(ctx, db, username, user.Spec.Roles)
```

Comment:
This needs to cater to situations where roles are modified to _remove_ permissions as well. See https://github.com/Azure/azure-service-operator/blob/main/v2/internal/util/mysql/privilege.go#L29 for what we do in MySQL. From my read of GrantUserRoles, it only adds permissions it never removes them

Reply:
I added logic to also revoke roles like in mysql and tested it with success. I also added a new option to set the WITH OPTIONS CREATEROLE, CREATEDB, LOGIN and REPLICATION.